source/extensions/filters/network/postgres_proxy/postgres_decoder.cc

```diff
@@ -176,134 +179,216 @@ void DecoderImpl::initialize() {
-  // Check if the entire message has been read.
-  std::string message;
-  message_len_ = data.peekBEInt<uint32_t>(startup_ ? 0 : 1);
-  if (data.length() < (message_len_ + (startup_ ? 0 : 1))) {
-    ENVOY_LOG(trace, "postgres_proxy: cannot parse message. Need {} bytes in buffer",
-              message_len_ + (startup_ ? 0 : 1));
-    // Not enough data in the buffer.
-    return Decoder::NeedMoreData;
+  if (validationResult == Message::ValidationFailed) {
+    // Message does not conform to the expected format. Move to out-of-sync state.
+    data.drain(data.length());
+    state_ = State::OutOfSyncState;
+    return Decoder::Result::ReadyForNext;
   }
 
-  if (startup_) {
-    uint32_t code = data.peekBEInt<uint32_t>(4);
-    // Startup message with 1234 in the most significant 16 bits
-    // indicate request to encrypt.
-    if (code >= 0x04d20000) {
-      encrypted_ = true;
-      // Handler for SSLRequest (Int32(80877103) = 0x04d2162f)
-      // See details in https://www.postgresql.org/docs/current/protocol-message-formats.html.
-      if (code == 0x04d2162f) {
-        // Notify the filter that `SSLRequest` message was decoded.
-        // If the filter returns true, it means to pass the message upstream
-        // to the server. If it returns false it means, that filter will try
-        // to terminate SSL session and SSLRequest should not be passed to the
-        // server.
-        encrypted_ = callbacks_->onSSLRequest();
-      }
-
-      // Count it as recognized frontend message.
-      callbacks_->incMessagesFrontend();
-      if (encrypted_) {
-        ENVOY_LOG(trace, "postgres_proxy: detected encrypted traffic.");
-        incSessionsEncrypted();
-        startup_ = false;
-      }
-      data.drain(data.length());
-      return encrypted_ ? Decoder::ReadyForNext : Decoder::Stopped;
+  Decoder::Result result = Decoder::Result::ReadyForNext;
+  uint32_t code = data.peekBEInt<uint32_t>(4);
+  data.drain(4);
+  // Startup message with 1234 in the most significant 16 bits
+  // indicate request to encrypt.
+  if (code >= 0x04d20000) {
+    encrypted_ = true;
+    // Handler for SSLRequest (Int32(80877103) = 0x04d2162f)
+    // See details in https://www.postgresql.org/docs/current/protocol-message-formats.html.
+    if (code == 0x04d2162f) {
+      // Notify the filter that `SSLRequest` message was decoded.
+      // If the filter returns true, it means to pass the message upstream
+      // to the server. If it returns false it means, that filter will try
+      // to terminate SSL session and SSLRequest should not be passed to the
+      // server.
+      encrypted_ = callbacks_->onSSLRequest();
+    }
+
+    // Count it as recognized frontend message.
+    callbacks_->incMessagesFrontend();
+    if (encrypted_) {
+      ENVOY_LOG(trace, "postgres_proxy: detected encrypted traffic.");
+      incSessionsEncrypted();
+      state_ = State::EncryptedState;
+      startup_ = false;
```

Comment:
Now with the new `state_` control seems this flag is useless, no?

Reply:
Yes. That is not used anymore. Thanks.